torch/_inductor/pattern_matcher.py

```diff
@@ -404,7 +580,117 @@ def visit(other_node):
         torch.fx.map_arg((node.args, node.kwargs), visit)
 
 
-def fx_passes(gm: torch.fx.GraphModule):
+def _not_implemented(*args, **kwargs):
+    raise NotImplementedError()
+
+
+def fx_to_pattern(gm, ignore_types=(), argnames=(), scalar_workaround=()):
+    """
+    Convert an FX graph into a PatternExpr.  This is useful for simple
+    patterns that can only match single functions and fixed length lists.
+    """
+    # scalar_workaround is a hack to capture dropout_p
+    # see https://github.com/pytorch/pytorch/issues/97894
+    scalar_workaround = scalar_workaround or {}
+    inv_scalar_workaround = {v: k for k, v in scalar_workaround.items()}
+    assert len(inv_scalar_workaround) == len(scalar_workaround)
+
+    def process_arg(x):
+        if isinstance(x, (float, int)) and x in inv_scalar_workaround:
+            return KeywordArg(inv_scalar_workaround[x])
+        if type(x) in ignore_types:
+            return Ignored()
+        return x
+
+    argnum = itertools.count()
+
+    class Converter(torch.fx.Interpreter):
+        call_method = _not_implemented
+        call_module = _not_implemented
+        get_attr = _not_implemented
+
+        def placeholder(self, target, args, kwargs):
+            n = next(argnum)
+            if n < len(argnames):
+                return KeywordArg(argnames[n])
+            if argnames:
+                assert target.startswith("tangent")
+                return KeywordArg(target)
+            else:
+                target = re.sub(r"_\d+$", "", target)  # de-mangle arg name
+                return KeywordArg(target)
+
+        def call_function(self, target, args, kwargs):
+            args, kwargs = pytree.tree_map(process_arg, (args, kwargs))
+            return CallFunction(target, *args, **kwargs)
+
+        def run_node(self, n):
+            rv = super().run_node(n)
+            rv.users = len(n.users)
+            return rv
+
+    pattern = Converter(gm).run()
+    if not isinstance(pattern, PatternExpr):
+        return MultiOutputPattern(pytree.tree_flatten(pattern)[0])
+    return pattern
+
+
+@torch.no_grad()
+def inference_graph(fn, args):
+    """Build a normalized inference graph, for use with fx_to_pattern"""
+    gm = make_fx(fn, select_decomp_table())(*args)
```

Comment:
Do we need to maintain any invariants here about functionalization ?

Reply:
We'll need to do the RNG functionalization at least I think right ?

Reply:
We will need the behavior here to match AotAutograd for the patterns being used.